### ISSUE TITLE
RewardStrings and Multi-Keywords

### ISSUE THREAD

**Issue by tass23** [Ravenwolfe/ServUO-XmlSpawner#1]
Arte mentioned a problem he was working on with the XmlQuest RewardString. Currently, an example rewardstring is built like this:
RewardString/@bag/Gold,200

This would reward the player with 200 gold when the quest has been completed/turned in.

The problem is that, everything after @bag is parsed into the RewardString. So if you wanted to put together multiple keywords, you can't, because they get tossed aside, or included in the RewardString, which breaks the reward. You are supposed to be able to use ; to separate keywords.

**Comment by Ravenwolfe** [comment]
Is this when reading from an xmlconfig file? Can you send me an example of the xml file you would be loading with multiple keywords?

**Comment by Ravenwolfe** [comment]
Also, I'm looking at the function to split the strings, and it appears they are split with a : not a ;

Have you tried that?

**Comment by tass23** [comment]
When reading from an XmlConfig file or when it's reading what was parsed into the entry book (on an Xmlspawner for example).

Here is an example:
`GIVE/<questholder/name/Trial of Flesh/titlestring/Trial of Flesh/notestring/Use Force Cleanse on yourself to remove the poison and use Jedi Grove to heal yourself if you've taken damage./Objective1/Use Force Cleanse on yourself./Description1/Use Force Cleanse on yourself./Objective2/Use Jedi Grove to heal yourself./Description2/Use Jedi Grove to heal yourself if needed./repeatable/false/Autoreward/true/CanSeeReward/false/Attachmentstring/XmlAddKarma,300/Rewardstring/@bag/ADD/ForceLightningDisc ; SETONTRIGMOB/POISON,3 ; SET,0x401195FF/Running/true`

The last thing that is parsed is the very last ADD/ForceLightningDisc. The ; is supposed to represent a new keyword being called, but they never "fire".

Just saw your comment on the split....mine is ; but ServUO is :

**Comment by Ravenwolfe** [comment]
Hmm, wonder why the change, have you tried the same config file on a clean ServUO?

**Comment by tass23** [comment]
In the above instance, more entries were added to facilitate using those last two commands that were needed. Alternatively, an outside Xmlspawner has been triggered to handle extra commands.

No, I have not tried one of my config files on ServUO, but I know sending ones that work for me, to other people (even changing the ; to : ), didn't work for them.

**Comment by Ravenwolfe** [comment]
That example is in the <Objects2> element of the xml file correct?

**Comment by Ravenwolfe** [comment]
Sorry, that should say its in the "<Objects2>" element correct?

**Comment by Ravenwolfe** [comment]
Damn this thing wont let me use symbols
Its in the Objects2 element of the xmlfile correct?

**Comment by Ravenwolfe** [comment]
This is really weird. I don't see anywhere that it is parsing for a keyword of "Rewardstring". I only see it parsing for a keyword of "Reward"

**Comment by Ravenwolfe** [comment]
Are you able to get into Discord?

**Comment by tass23** [comment]
I think it's the `@` symbol specifically that is the key parsing character.

The above example is in an ACTION for an XmlDialog.

**Comment by Ravenwolfe** [comment]
@ symbol is usually a string literal. I see the parsing for Autoreward, CanSeeReward, etc. Just nothing for the Rewardstring as a keyword. Very odd.

**Comment by Ravenwolfe** [comment]
valid_entry = true;
							boolEntry = false;
							try { boolEntry = bool.Parse((string)dr["AutoReward"]); }
							catch { valid_entry = false; }
							if (valid_entry)
							{
								this.AutoReward = boolEntry;
							}

							valid_entry = true;
							boolEntry = true;
							try { boolEntry = bool.Parse((string)dr["CanSeeReward"]); }
							catch { valid_entry = false; }
							if (valid_entry)
							{
								this.CanSeeReward = boolEntry;
							}

**Comment by Ravenwolfe** [comment]
Clearly, its parsing the rewardstring somewhere, since I can see it wants to reward a bag.

**Comment by tass23** [comment]
Arte said this:

> you need to use the @ operator so that the literal string "bag/ADD/expstonefifty/ADD/expstonefifty" is assigned as the reward string. If you didnt have the @, then it would try to parse the special '/' characters. The @ allows you to include such special chars in your rewardstring.

**Comment by tass23** [comment]
So, does that mean the keywords should be marked as "special characters" instead??
Problem solved...?

**Comment by Ravenwolfe** [comment]
Exactly, the @ tells the system that everything after it should be treated as a literal string. Otherwise it would not know that the "/" is meant to be read as a / instead of a special character. This is why everything after it is being read as part of the rewardstring. But there is still an issue. It should be parsing based on the : symbol BEFORE it reads the rewardstring, which should keep anything after the : from being read as a string. I still cant find where it parses for the keyword "rewardstring" tho. Can you try adding a : (or in your case a ; ) before and after rewardstring?

**Comment by Ravenwolfe** [comment]
I also noticed this:
				// if the reward item is not set, and the reward string is specified, then use the reward string to construct and assign the
				// reward item
				// dont allow player made quests to use the rewardstring creation feature

**Comment by Ravenwolfe** [comment]
Sounds like playermade quests are not meant to use rewardstring??

**Comment by tass23** [comment]
This isn't a player made quest though. The way I learned to create quests was how Arte originally explained it. As for the player made quests, I have never used that part of the system. Unless the bug carries over on both sides...?

In either case though, using : or ; should terminate that entry and allow usage of the new keyword on that line.

**Comment by Ravenwolfe** [comment]
Yeah, and it do the splits using ":" before the keywords are read. Or at least, it should be.

**Comment by Ravenwolfe** [comment]
And I can see where is is splitting using the "/". But I still dont see anywhere using "rewardstring" as a keyword.

**Comment by Ravenwolfe** [comment]
It doesnt exist in ANY keyword list.

**Comment by Ravenwolfe** [comment]
The only place I can see RewardString getting populated is here:

							valid_entry = true;
							strEntry = null;
							try { strEntry = (string)dr["Reward"]; }
							catch { valid_entry = false; }
							if (valid_entry)
							{
								this.RewardString = strEntry;
							}

**Comment by Ravenwolfe** [comment]
And that is looking to parse the element "Reward" from an xml file. Like this example:

<XmlQuestToken>
	<Properties>
		<Name>The Quest</Name>
		<Title>Do some quest stuff</Title>
		<PartyEnabled>true</PartyEnabled>
		<PartyRange>30</PartyRange>
		<Objective1>KILL,balron,10</Objective1>
		<Objective2>COLLECT,ginseng,100</Objective2>
		<Objective3>Go talk to the man</Objective3>
		<Objective4>KILL,harpy,5</Objective4>
		<Objective5>Return to collect your reward</Objective5>
		<Expiration>20</Expiration>
		<Reward>GIVE/&lt;longsword/name/Blazer/weaponattributes.hitfireball/RND,30,70&gt;</Reward>
	</Properties>
</XmlQuestToken>

**Comment by tass23** [comment]
AttachmentString is handled in a similar manner. But you're not incorporating the `@` in those cases, because you're changing an attachment value.

So is it just the `@` that is being recognized as the reward string, like a constructor?

**Comment by Ravenwolfe** [comment]
No, the @ just means treat the following as a literal string. As an example..

string = @"all of this is" text

will = "all of this is" text

without the literal it would be this:

string = "all of this is" text

would only =  all of this is

**Comment by Ravenwolfe** [comment]
The @ means treat everything after as one long string and ignore any special characters.

**Comment by Ravenwolfe** [comment]
Verbatim string literals start with @ and are also enclosed in double quotation marks. For example:
C#

@"good morning"  // a string literal  

The advantage of verbatim strings is that escape sequences are not processed, which makes it easy to write, for example, a fully qualified file name:
C#

@"c:\Docs\Source\a.txt"  // rather than "c:\\Docs\\Source\\a.txt"

**Comment by tass23** [comment]
Okay, so the function of `@` needs to be altered, to allow for those breaks in line code, or the keywords need to supercede it, so multiple keywords can be strung together. That would work for the XmlDialog, but not on an Xmlspawner, because you aren't utilizing `@` in those cases. Multiple keywords must be split into individual entries and then grouped or the Xmlspawner set as sequential.

**Comment by Ravenwolfe** [comment]
Try it like this:

GIVE/<questholder/name/Trial of Flesh/titlestring/Trial of Flesh/notestring/Use Force Cleanse on yourself to remove the poison and use Jedi Grove to heal yourself if you've taken damage./Objective1/Use Force Cleanse on yourself./Description1/Use Force Cleanse on yourself./Objective2/Use Jedi Grove to heal yourself./Description2/Use Jedi Grove to heal yourself if needed./repeatable/false/Autoreward/true/CanSeeReward/false/Attachmentstring/XmlAddKarma,300/Rewardstring/@"bag/ADD/ForceLightningDisc" ; SETONTRIGMOB/POISON,3 ; SET,0x401195FF/Running/true

**Comment by Ravenwolfe** [comment]
We can't change the function of the @ symbol, that is a function in C# that is built in.

**Comment by Ravenwolfe** [comment]
What I really need to figure out, is how is rewardstring being triggered as a keyword. Once I can figure that out, we should be able to fix the issue.

**Comment by tass23** [comment]
I'll try that code you posted when I get back home and let you know.
RewardString, being the keyword, then has constructors, so let's just change the constructors...?

**Comment by tass23** [comment]
Using "" instead of `@` didn't work. There was an error visible in the questholder:
invalid type specification: "bag

There is another thing I would like to put on the table for consideration; the ability to add a keyword into the Rewardstring. For example, when a player completes a quest and the AutoReward/True, and the Rewardstring is set @bag/Gold,200
Right after the Rewardstring, I would like to do this:
Rewardstring/Set,SpawnerName/DoReset/True/Running/True/@bag/Gold,200

The reason for this is that on completion of the quest, that keyword would be executed. Whereas, adding them after the Rewardstring as a new command using ; or : gets executed right after the questholder is given to the player. Perhaps a new string type, like KeywordString, that can get executed with the Rewardstring. I'm thinking this code could be used to create public XmlCommand KeywordString

```
public Item RewardItem
        {
            get
            {
                // if the reward item is not set, and the reward string is specified, then use the reward string to construct and assign the
                // reward item
                // dont allow player made quests to use the rewardstring creation feature
                if ((m_RewardItem == null || m_RewardItem.Deleted) &&
                    (m_RewardString != null) && !PlayerMade)
                {
                    object o = XmlQuest.CreateItem(this, m_RewardString, out m_status_str, typeof(Item));
                    if (o is Item)
                    {
                        m_RewardItem = o as Item;
                        PackItem(m_RewardItem);
                    }
                    else
                        if (o is XmlAttachment)
                        {
                            ((XmlAttachment)o).Delete();
                        }
                }

                return m_RewardItem;
            }
            set
            {
                // get rid of any existing reward item if it has been assigned
                if (m_RewardItem != null && !m_RewardItem.Deleted)
                {

                    ReturnReward();
                }

                // and assign the new item
                m_RewardItem = value;

                /*
                // is this currently carried by a mobile?
                if(m_RewardItem.RootParent != null && m_RewardItem.RootParent is Mobile)
                {
                    // if so then remove it
                    ((Mobile)(m_RewardItem.RootParent)).RemoveItem(m_RewardItem);

                }
                */

                // and put it in the pack
                if (m_RewardItem != null && !m_RewardItem.Deleted)
                {
                    PackItem(m_RewardItem);
                }


            }
        }
```
